lib/execution_engine2/execution_engine2Impl.py

```diff
@@ -375,6 +375,81 @@ def run_job_batch(self, ctx, params, batch_params):
         # return the results
         return [job_ids]
 
+    def retry_job(self, ctx, params):
+        """
+        #TODO write retry parent tests to ensure BOTH the parent_job_id is present, and retry_job_id is present
+        #TODO Add retry child that checks the status of the child? to prevent multiple retries
+        Allowed Jobs
+         Regular Job with no children
+         Regular job with/without parent_id that runs a kbparallel call or a run_job_batch call
+        Not Allowed
+         Regular Job with children (Should not be possible to create yet)
+         Batch Job Parent Container (Not a job, it won't do anything, except cancel it's child jobs)
+        :param params: instance of type "RetryParams" (job_id of job to retry
+           as_admin: retry someone elses job in your namespace #TODO Possibly
+           Add JobRequirements job_requirements;) -> structure: parameter
+           "job_id" of type "job_id" (A job id.), parameter "as_admin" of
+           type "boolean" (@range [0,1])
+        :returns: instance of type "RetryResult" (job_id of retried job
+           retry_id: job_id of the job that was launched str error: reason as
+           to why that particular retry failed (available for bulk retry
+           only)) -> structure: parameter "job_id" of type "job_id" (A job
+           id.), parameter "retry_id" of type "job_id" (A job id.), parameter
+           "error" of String
+        """
+        # ctx is the context object
+        # return variables are: retry_result
+        #BEGIN retry_job
+        mr = SDKMethodRunner(
+            user_clients=self.gen_cfg.get_user_clients(ctx),
+            clients = self.clients,
+            job_permission_cache=self.job_permission_cache,
+            admin_permissions_cache=self.admin_permissions_cache
+        )
+        retry_result = mr.retry(job_id=params['job_id'], as_admin=params.get('as_admin'))
```

Comment:
I'm still pretty strongly against throwing a `KeyError` when missing parameters - it can be really difficult for the client to figure out what's wrong, especially in KBase land where often the exception message is displayed without the exception type. I would switch to a `get()` and have `EE2Runjob` check for falsy values and throw a more understandable error.

Reply:
changed to get. errors are thrown further down the code now, once in retry_multiple, once in MongoUtil `raise ValueError("Please provide a valid job id")`

Reply:
:+1: